<!doctype html>
<!--
  Minimal Mistakes Jekyll Theme 4.24.0 by Michael Rose
  Copyright 2013-2020 Michael Rose - mademistakes.com | @mmistakes
  Free for personal and commercial use under the MIT license
  https://github.com/mmistakes/minimal-mistakes/blob/master/LICENSE
-->
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">

<!-- begin _includes/seo.html --><title>Lesson 2 - Tree structure - Biogeography and Macroecology 2023</title>
<meta name="description" content="">



<meta property="og:type" content="article">
<meta property="og:locale" content="en_US">
<meta property="og:site_name" content="Biogeography and Macroecology 2023">
<meta property="og:title" content="Lesson 2 - Tree structure">
<meta property="og:url" content="http://localhost:4000/updates/02_Tree-structure/">




  <meta property="og:image" content="http://localhost:4000/assets/images/tulane.jpg">



  <meta name="twitter:site" content="@Barnabas_Daru">
  <meta name="twitter:title" content="Lesson 2 - Tree structure">
  <meta name="twitter:description" content="">
  <meta name="twitter:url" content="http://localhost:4000/updates/02_Tree-structure/">

  
    <meta name="twitter:card" content="summary_large_image">
    <meta name="twitter:image" content="http://localhost:4000/assets/images/tulane.jpg">
  

  



  <meta property="article:published_time" content="2023-12-08T00:00:00-06:00">





  

  


<link rel="canonical" href="http://localhost:4000/updates/02_Tree-structure/">




<script type="application/ld+json">
  {
    "@context": "https://schema.org",
    
      "@type": "Person",
      "name": null,
      "url": "http://localhost:4000/",
      "sameAs": ["https://github.com/darunabas/phyloregion-course-2022","https://twitter.com/Barnabas_Daru"]
    
  }
</script>







<!-- end _includes/seo.html -->



  <link href="/feed.xml" type="application/atom+xml" rel="alternate" title="Biogeography and Macroecology 2023 Feed">


<!-- https://t.co/dKP3o1e -->
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<script>
  document.documentElement.className = document.documentElement.className.replace(/\bno-js\b/g, '') + ' js ';
</script>

<!-- For all browsers -->
<link rel="stylesheet" href="/assets/css/main.css">
<link rel="preload" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5/css/all.min.css" as="style" onload="this.onload=null;this.rel='stylesheet'">
<noscript><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5/css/all.min.css"></noscript>



    <!-- start custom head snippets -->

<!-- insert favicons. use https://realfavicongenerator.net/ -->

<!-- end custom head snippets -->

  </head>

  <body class="layout--posts wide">
    <nav class="skip-links">
  <ul>
    <li><a href="#site-nav" class="screen-reader-shortcut">Skip to primary navigation</a></li>
    <li><a href="#main" class="screen-reader-shortcut">Skip to content</a></li>
    <li><a href="#footer" class="screen-reader-shortcut">Skip to footer</a></li>
  </ul>
</nav>

    <!--[if lt IE 9]>
<div class="notice--danger align-center" style="margin: 0;">You are using an <strong>outdated</strong> browser. Please <a href="https://browsehappy.com/">upgrade your browser</a> to improve your experience.</div>
<![endif]-->

    

<div class="masthead">
  <div class="masthead__inner-wrap">
    <div class="masthead__menu">
      <nav id="site-nav" class="greedy-nav">
        
        <a class="site-title" href="/">
          Biogeography and Macroecology 2023
          
        </a>
        <ul class="visible-links"><li class="masthead__menu-item">
              <a href="/team/">Our Team</a>
            </li><li class="masthead__menu-item">
              <a href="/schedule/">Schedule</a>
            </li><li class="masthead__menu-item">
              <a href="/blog/">Content</a>
            </li><li class="masthead__menu-item">
              <a href="/conduct/">Conduct</a>
            </li><li class="masthead__menu-item">
              <a href="/resources/">Resources</a>
            </li></ul>
        
        <button class="greedy-nav__toggle hidden" type="button">
          <span class="visually-hidden">Toggle menu</span>
          <div class="navicon"></div>
        </button>
        <ul class="hidden-links hidden"></ul>
      </nav>
    </div>
  </div>
</div>


    <div class="initial-content">
      
  







<div class="page__hero--overlay"
  style=" background-image: linear-gradient(rgba(0, 0, 0, 0.2), rgba(0, 0, 0, 0.2)), url('/assets/images/tulane.jpg');"
>
  
    <div class="wrapper">
      <h1 id="page-title" class="page__title" itemprop="headline">
        
          Lesson 2 - Tree structure

        
      </h1>
      
        <p class="page__lead">
</p>
      
      

  <p class="page__meta">
    

    

    
      
      

      <span class="page__meta-readtime">
        <i class="far fa-clock" aria-hidden="true"></i>
        
          10 minute read
        
      </span>
    
  </p>


      
      
    </div>
  
  
</div>





<div id="main" role="main">
  


  <div class="archive">
    
    <!-- Look the author details up from the site config. -->


<!-- Output author details if some exist. -->


<h2 id="lecture-2">Lecture 2</h2>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">knitr</span><span class="o">::</span><span class="n">opts_chunk</span><span class="o">$</span><span class="n">set</span><span class="p">(</span><span class="n">echo</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">knitr</span><span class="o">::</span><span class="n">opts_chunk</span><span class="o">$</span><span class="n">set</span><span class="p">(</span><span class="n">class.source</span><span class="o">=</span><span class="s1">'fold-show'</span><span class="p">)</span><span class="w">
</span><span class="n">knitr</span><span class="o">::</span><span class="n">opts_chunk</span><span class="o">$</span><span class="n">set</span><span class="p">(</span><span class="n">collapse</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">phangorn</span><span class="p">)</span><span class="w">
</span><span class="n">nr_desc</span><span class="w"> </span><span class="o">&lt;-</span><span class="w">  </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">){</span><span class="w">
    </span><span class="n">x</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">reorder</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="s2">"postorder"</span><span class="p">)</span><span class="w">
    </span><span class="n">res</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">numeric</span><span class="p">(</span><span class="nf">max</span><span class="p">(</span><span class="n">x</span><span class="o">$</span><span class="n">edge</span><span class="p">))</span><span class="w">
    </span><span class="n">res</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="n">Ntip</span><span class="p">(</span><span class="n">x</span><span class="p">)]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1L</span><span class="w">
    </span><span class="k">for</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">nrow</span><span class="p">(</span><span class="n">x</span><span class="o">$</span><span class="n">edge</span><span class="p">)){</span><span class="w">
        </span><span class="n">tmp</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">x</span><span class="o">$</span><span class="n">edge</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="m">1</span><span class="p">]</span><span class="w">
        </span><span class="n">res</span><span class="p">[</span><span class="n">tmp</span><span class="p">]</span><span class="w"> </span><span class="o">=</span><span class="w">  </span><span class="n">res</span><span class="p">[</span><span class="n">tmp</span><span class="p">]</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">res</span><span class="p">[</span><span class="n">x</span><span class="o">$</span><span class="n">edge</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="p">]</span><span class="w">
    </span><span class="p">}</span><span class="w">
    </span><span class="n">res</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="n">root_to_tip</span><span class="w"> </span><span class="o">&lt;-</span><span class="w">  </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">){</span><span class="w">
    </span><span class="n">x</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">reorder</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w">
    </span><span class="n">res</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">numeric</span><span class="p">(</span><span class="nf">max</span><span class="p">(</span><span class="n">x</span><span class="o">$</span><span class="n">edge</span><span class="p">))</span><span class="w">
    </span><span class="k">for</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">nrow</span><span class="p">(</span><span class="n">x</span><span class="o">$</span><span class="n">edge</span><span class="p">)){</span><span class="w">
        </span><span class="n">pa</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">x</span><span class="o">$</span><span class="n">edge</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="m">1</span><span class="p">]</span><span class="w">
        </span><span class="n">ch</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">x</span><span class="o">$</span><span class="n">edge</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="m">2</span><span class="p">]</span><span class="w">        
        </span><span class="n">res</span><span class="p">[</span><span class="n">ch</span><span class="p">]</span><span class="w"> </span><span class="o">=</span><span class="w">  </span><span class="n">res</span><span class="p">[</span><span class="n">pa</span><span class="p">]</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">x</span><span class="o">$</span><span class="n">edge.length</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w">
    </span><span class="p">}</span><span class="w">
    </span><span class="n">res</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre></div></div>

<h2 id="concepts-today">Concepts today</h2>

<ul>
  <li>How is a phylogenetic tree stored in R?</li>
  <li>How can we traverse a tree to see and manipulate nodes and branches?</li>
  <li>What are some major packages and key functions in the R computing environment?</li>
  <li>Beyond the tree: What are some important packages that implement comparative methods, biogeography, and visualization?</li>
</ul>

<h2 id="a-brief-history-of-phylogenetics-in-r">A brief history of phylogenetics in R</h2>

<p>1993: first release of R</p>

<p>1996: Ihaka &amp; Gentleman, J. Comput. Graph. Statist., 5: 299–314</p>

<p>2000: R 1.0.0</p>

<p>2001: start the development of ape</p>

<p>2002: first release of ape</p>

<p>2008: first release of phangorn</p>

<p>2019: 260 packages depending on ape on CRAN, 16 on BioConductor (and 100’s elsewhere: R-Forge, GitHub, . . .)</p>

<h2 id="getting-help">Getting help</h2>

<p>So many functions, but how to find/remember them?</p>

<ul>
  <li>write scripts (and maintain them!)</li>
  <li>help.start()</li>
  <li>?… (or help(“…”))</li>
  <li>apropos(“…”)</li>
  <li>use TAB-completion (or a “cool” editor)</li>
</ul>

<p>Operators are also functions:</p>
<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">`+`</span><span class="w"> </span><span class="c1"># or get("+")</span><span class="w">
</span><span class="n">`+`</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>Find generic functions for a class</p>

<p>Common generic functions are print, plot, c, summary, anova, [, …</p>
<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">ape</span><span class="p">)</span><span class="w">
</span><span class="n">methods</span><span class="p">(</span><span class="n">class</span><span class="o">=</span><span class="s2">"phylo"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>
<p>Run all examples of the help files!</p>
<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">example</span><span class="p">(</span><span class="n">plot.phylo</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>
<p>However examples in (CRAN) R packages are not allowed to run longer than 5 seconds. So for inference this results frequently in toy examples. You need to change to higher number of bootstrap iterations, change stop criterion for optimisation et cetera.</p>

<p>Many packages on CRAN and all on Bioconductor come with vignettes.</p>
<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">vignette</span><span class="p">()</span><span class="w">
</span><span class="n">vignette</span><span class="p">(</span><span class="n">package</span><span class="o">=</span><span class="s2">"phangorn"</span><span class="p">)</span><span class="w">
</span><span class="n">vignette</span><span class="p">(</span><span class="s2">"Trees"</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="o">=</span><span class="s2">"phangorn"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>And there is google and stackoverflow.</p>

<h2 id="the-phylo-class">The <code class="language-plaintext highlighter-rouge">phylo</code> class</h2>

<p>The <code class="language-plaintext highlighter-rouge">phylo</code> class is a tree structure in <code class="language-plaintext highlighter-rouge">ape</code> are the standard for storing phylogenetic in R. It is also used in more than 260 packages on CRAN.</p>

<p>This class represents phylogenetic trees and networks; (internal) nodes are of
degree ≥ 2; the tips (terminal nodes) are of degree 1. The tip labels may be
replicated (or can even missing). Idem for the node labels (which are optional).</p>

<h2 id="structure-of-phylo-objects">Structure of phylo objects</h2>

<p>An object of class <code class="language-plaintext highlighter-rouge">phylo</code> is a list containing at least the following objects:</p>

<ol>
  <li>An <em>integer</em> matrix named edge with two columns and as many rows as there are edges in the tree.</li>
  <li>A <em>character</em> vector named <code class="language-plaintext highlighter-rouge">tip.label</code>, can be “”.</li>
  <li>An <em>integer</em> <code class="language-plaintext highlighter-rouge">Nnode</code> with the number of internal nodes.</li>
  <li>A class attribute <code class="language-plaintext highlighter-rouge">phylo</code>.</li>
</ol>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">ape</span><span class="p">)</span><span class="w">
</span><span class="n">tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.tree</span><span class="p">(</span><span class="n">text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"(,(,));"</span><span class="p">)</span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span><span class="w">
</span><span class="n">nodelabels</span><span class="p">()</span><span class="w">
</span><span class="n">tiplabels</span><span class="p">()</span><span class="w">
</span><span class="n">str</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>
<p>The storage mode is important. Saving a vector of integers saves memory in comparison to vector of doubles. 
Also objects of class <code class="language-plaintext highlighter-rouge">phylo</code> often in C-Code so you need to take care of this.</p>

<p>Also comparing doubles is dangerous:</p>
<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">x</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">3</span><span class="w">
</span><span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sqrt</span><span class="p">(</span><span class="m">3</span><span class="p">)</span><span class="o">^</span><span class="m">2</span><span class="w">
</span><span class="n">x</span><span class="o">==</span><span class="n">y</span><span class="w">
</span></code></pre></div></div>

<p>In the edge matrix each edge (branch) is coded by the nodes it connects.</p>
<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">tree</span><span class="o">$</span><span class="n">edge</span><span class="w">
</span></code></pre></div></div>
<p>Tips (terminal nodes, leaves, labels, …) are numbered from 1 to <code class="language-plaintext highlighter-rouge">n</code>, where <code class="language-plaintext highlighter-rouge">n</code> is the number of tips. The root is numbered <code class="language-plaintext highlighter-rouge">n+1</code>. The ancestral node (closer to the root) is always on the lefthand column.</p>

<p>Some important properties and observations:</p>
<ul>
  <li>the first column only contains values greater than <code class="language-plaintext highlighter-rouge">n</code>.</li>
  <li>all nodes, but the root, appear exactly once in the second column.</li>
  <li>in a strictly bifurcating tree, all nodes greater then <code class="language-plaintext highlighter-rouge">n</code> appear twice and the root node <code class="language-plaintext highlighter-rouge">n+1</code>, in case of ‘twice’ for rooted and <code class="language-plaintext highlighter-rouge">trice</code> for unrooted trees.</li>
</ul>

<p>So we can start and construct a minimal tree ourselves:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">ape</span><span class="p">)</span><span class="w">
</span><span class="n">simple_tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">structure</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">edge</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">3L</span><span class="p">,</span><span class="w"> </span><span class="m">3L</span><span class="p">,</span><span class="w"> </span><span class="m">1L</span><span class="p">,</span><span class="w"> </span><span class="m">2L</span><span class="p">),</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w"> 
                        </span><span class="n">tip.label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"t1"</span><span class="p">,</span><span class="w"> </span><span class="s2">"t2"</span><span class="p">),</span><span class="w"> </span><span class="n">Nnode</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1L</span><span class="p">),</span><span class="w"> </span><span class="n">class</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"phylo"</span><span class="p">)</span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="n">simple_tree</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>A <code class="language-plaintext highlighter-rouge">phylo</code> object always contains an <code class="language-plaintext highlighter-rouge">edge</code>, <code class="language-plaintext highlighter-rouge">tip.label</code> and <code class="language-plaintext highlighter-rouge">Nnode</code>. Often a phylo object contains some of the following elements:</p>

<ol>
  <li>a <em>numeric</em> vector <code class="language-plaintext highlighter-rouge">edge.length</code>. The length of the vector is number of rows of the <code class="language-plaintext highlighter-rouge">edge</code> matrix and the order corresponds to the rows in <code class="language-plaintext highlighter-rouge">edge</code>.</li>
  <li>a <em>character</em> vector <code class="language-plaintext highlighter-rouge">node.labels</code> of length <code class="language-plaintext highlighter-rouge">Nnode</code>. The node labels are labeled from <code class="language-plaintext highlighter-rouge">n+1</code> to <code class="language-plaintext highlighter-rouge">n+Nnode</code></li>
  <li>A single <em>numeric</em> root.edge providing the edge length at the root.</li>
</ol>

<p>Additionally most <code class="language-plaintext highlighter-rouge">phylo</code> objects have an attribute <code class="language-plaintext highlighter-rouge">order</code>, which can be either “cladewise”, “postorder” or “pruningwise”.</p>

<h2 id="tree-traversal">Tree traversal</h2>

<p>Tree traversal in ape is implemented iterative in contrast to the recursive implementation in most other phylogenetic programs.</p>

<h3 id="preorder-tree-traversal">Preorder tree traversal</h3>
<p><em>Preorder tree traversal</em> is travelling through the tree from the root to the tips visiting always the all nodes the right subtree before visiting the right subtree.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rtree</span><span class="p">(</span><span class="m">5</span><span class="p">)</span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span><span class="w"> </span><span class="n">show.tip.label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span><span class="n">nodelabels</span><span class="p">()</span><span class="w">
</span><span class="n">tiplabels</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">reorder</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span><span class="w"> </span><span class="s2">"cladewise"</span><span class="p">)</span><span class="o">$</span><span class="n">edge</span><span class="w">
</span></code></pre></div></div>

<h3 id="postorder-tree-traversal">Postorder tree traversal</h3>
<p><em>Postorder tree traversal</em> travels through the tree from the tips through the root. All the descendent nodes need to be visited before the ancestral node.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">reorder</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span><span class="w"> </span><span class="s2">"postorder"</span><span class="p">)</span><span class="o">$</span><span class="n">edge</span><span class="w">
</span></code></pre></div></div>

<p>Defining the order of the traversal of the tree this way allows to iterate just through the tree. Reorder not only changes the order of the <code class="language-plaintext highlighter-rouge">edge</code> matrix, but also of the <code class="language-plaintext highlighter-rouge">edge.length</code> if it exists.</p>

<p>###Excercise 1:
Write a function which counts how many descendants each node has. 
The output should look like:</p>
<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">set.seed</span><span class="p">(</span><span class="m">42</span><span class="p">)</span><span class="w">
</span><span class="n">tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rtree</span><span class="p">(</span><span class="m">5</span><span class="p">)</span><span class="w">

</span><span class="n">plot</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span><span class="w"> </span><span class="n">show.tip.label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span><span class="n">nodelabels</span><span class="p">()</span><span class="w">
</span><span class="n">tiplabels</span><span class="p">()</span><span class="w">
</span><span class="n">nr_desc</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>My solutions</p>
<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">phangorn</span><span class="p">)</span><span class="w">
</span><span class="n">nr_desc_cheat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">){</span><span class="w">
  </span><span class="n">lengths</span><span class="p">(</span><span class="n">Descendants</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="nf">max</span><span class="p">(</span><span class="n">x</span><span class="o">$</span><span class="n">edge</span><span class="p">)))</span><span class="w">
</span><span class="p">}</span><span class="w"> 


</span><span class="n">nr_desc</span><span class="w"> </span><span class="o">&lt;-</span><span class="w">  </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">){</span><span class="w">
    </span><span class="n">x</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">reorder</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="s2">"postorder"</span><span class="p">)</span><span class="w">
    </span><span class="n">res</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">numeric</span><span class="p">(</span><span class="nf">max</span><span class="p">(</span><span class="n">x</span><span class="o">$</span><span class="n">edge</span><span class="p">))</span><span class="w">
    </span><span class="n">res</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="n">Ntip</span><span class="p">(</span><span class="n">x</span><span class="p">)]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1L</span><span class="w">
    </span><span class="k">for</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">nrow</span><span class="p">(</span><span class="n">x</span><span class="o">$</span><span class="n">edge</span><span class="p">)){</span><span class="w">
        </span><span class="n">tmp</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">x</span><span class="o">$</span><span class="n">edge</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="m">1</span><span class="p">]</span><span class="w">
        </span><span class="n">res</span><span class="p">[</span><span class="n">tmp</span><span class="p">]</span><span class="w"> </span><span class="o">=</span><span class="w">  </span><span class="n">res</span><span class="p">[</span><span class="n">tmp</span><span class="p">]</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">res</span><span class="p">[</span><span class="n">x</span><span class="o">$</span><span class="n">edge</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="p">]</span><span class="w">
    </span><span class="p">}</span><span class="w">
    </span><span class="n">res</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre></div></div>

<p>###Excercise 2:
Write a function which computes the distance from each node to the root.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">tree</span><span class="o">$</span><span class="n">edge.length</span><span class="p">[]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1</span><span class="w">
</span><span class="n">root_to_tip</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>My solution:
```{r class.source = NULL}
root_to_tip &lt;-  function(x){
    x &lt;- reorder(x)
    res &lt;- numeric(max(x$edge))
    for(i in 1:nrow(x$edge)){
        pa &lt;- x$edge[i,1]
        ch &lt;- x$edge[i,2]      <br />
        res[ch] =  res[pa] + x$edge.length[i]
    }
    res
}</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>

## The 'multiPhylo' class

An object of class 'multiPhylo' is a list of of several trees each of class 
'phylo'. If all trees share the same tip labels, as in bootstrap samples or from
MCMC analyses, trees can be saved in a memory efficient way storing the labels
only once in an attribute *TipLabel*. 
 
```r
trees &lt;- rmtree(1000, 100)
trees
attr(trees, "TipLabel")
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">trees_compact</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">.compressTipLabel</span><span class="p">(</span><span class="n">trees</span><span class="p">)</span><span class="w">
</span><span class="nf">attr</span><span class="p">(</span><span class="n">trees_compact</span><span class="p">,</span><span class="w"> </span><span class="s2">"TipLabel"</span><span class="p">)</span><span class="w">

</span><span class="n">object.size</span><span class="p">(</span><span class="n">trees</span><span class="p">)</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">object.size</span><span class="p">(</span><span class="n">trees_compact</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>Switching between the two representations can be done via ‘.compressTipLabel’ and ‘.uncompressTipLabel’.</p>

<h2 id="importing-and-exporting-trees-and-networks">Importing and exporting trees and networks</h2>

<p>Package <code class="language-plaintext highlighter-rouge">ape</code> and <code class="language-plaintext highlighter-rouge">phangorn</code> contains functions to import trees and splits
networks:</p>

<ul>
  <li><code class="language-plaintext highlighter-rouge">read.tree</code> Newick files</li>
  <li><code class="language-plaintext highlighter-rouge">read.nexus</code> NEXUS files (only trees)</li>
  <li><code class="language-plaintext highlighter-rouge">read.evonet</code> networks in extended Newick format</li>
</ul>

<p>and</p>

<ul>
  <li><code class="language-plaintext highlighter-rouge">write.tree</code> Newick (one or several trees)</li>
  <li><code class="language-plaintext highlighter-rouge">write.nexus</code> NEXUS (one or several trees with or without translate block)</li>
  <li><code class="language-plaintext highlighter-rouge">write.evonet</code> extended Newick (.nhx)</li>
  <li><code class="language-plaintext highlighter-rouge">read.networx</code> phylogenetic splits networks in nexus format (e.g. Splitstree)</li>
</ul>

<h3 id="example">Example:</h3>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">write.tree</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span><span class="w"> </span><span class="s2">"example_tree.nex"</span><span class="p">)</span><span class="w"> 
</span><span class="n">tree_read</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.tree</span><span class="p">(</span><span class="n">file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"example_tree.nex"</span><span class="p">)</span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="n">tree_read</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>Any R object(s) can be saved on disk: this is much better if these data are used
only in R. (*Faster to read/write, smaller files, and no loss of numerical accuracy.)
Several objects:</p>

<ul>
  <li><code class="language-plaintext highlighter-rouge">save(X, a, y, file = "X.rda")</code> several objects with their names</li>
  <li>
    <p><code class="language-plaintext highlighter-rouge">load("X.rda")</code> $\rightarrow$ restores object names (i.e., possibly deletes previous
X, a, y in the workspace)
A single object:</p>
  </li>
  <li><code class="language-plaintext highlighter-rouge">saveRDS(X, file = "X.rds")</code></li>
  <li><code class="language-plaintext highlighter-rouge">Y &lt;- readRDS("X.rds")</code> $\rightarrow$ the name of the object can be changed</li>
</ul>

<h2 id="functions-for-phylo-objects">Functions for <code class="language-plaintext highlighter-rouge">phylo</code> objects</h2>

<p>Before you write a function look what’s already out there.</p>

<h3 id="some-convenience-functions">Some convenience functions</h3>
<p>There are several convenience functions to achieve common tasks:</p>

<table>
  <thead>
    <tr>
      <th>Operation</th>
      <th>High level</th>
      <th>Low level</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>How many tips?</td>
      <td>Ntip(tr)</td>
      <td>length(tr$tip.label)</td>
    </tr>
    <tr>
      <td>How many node?</td>
      <td>Nnode(tr)</td>
      <td>tr$Nnode</td>
    </tr>
    <tr>
      <td>How many edges?</td>
      <td>Nedge(tr)</td>
      <td>nrow(tr$edge)</td>
    </tr>
    <tr>
      <td>What are the ancestor of x</td>
      <td>Ancestors(tr, x, “parent”)</td>
      <td>tr$edge[tr$edge[,2]==x , 1]</td>
    </tr>
    <tr>
      <td>What are the descendants of x</td>
      <td>Descendants(tr, x, “children”)</td>
      <td>tr$edge[tr$edge[,1]==x , 2]</td>
    </tr>
  </tbody>
</table>

<p>Ntip, Nnode, Nedge are generic and work on <code class="language-plaintext highlighter-rouge">phylo</code> and <code class="language-plaintext highlighter-rouge">multiPhylo</code> objects. <code class="language-plaintext highlighter-rouge">Ancestors</code> and <code class="language-plaintext highlighter-rouge">Descendants</code> are vectorized , i.e. argument node can be a vector.</p>

<h2 id="some-useful-functions-to-modify-trees">Some useful functions to modify trees</h2>
<p><code class="language-plaintext highlighter-rouge">root(phy, outgroup, ...)</code>, 
<code class="language-plaintext highlighter-rouge">unroot(phy)</code>, 
<code class="language-plaintext highlighter-rouge">midpoint(tree)</code>,
<code class="language-plaintext highlighter-rouge">drop.tip(phy, tip, ...)</code>, 
<code class="language-plaintext highlighter-rouge">keep.tip(phy, tip)</code>, 
<code class="language-plaintext highlighter-rouge">extract.clade(phy, node, ...)</code>, 
<code class="language-plaintext highlighter-rouge">bind.tree(x, y, ...)</code>,
<code class="language-plaintext highlighter-rouge">c(...)</code>, 
<code class="language-plaintext highlighter-rouge">multi2di(phy, random = TRUE)</code>, 
<code class="language-plaintext highlighter-rouge">di2multi(phy, tol = 1e-8)</code>, 
<code class="language-plaintext highlighter-rouge">rotate(phy, node)</code>, 
<code class="language-plaintext highlighter-rouge">rotateConstr(phy, X)</code>, 
<code class="language-plaintext highlighter-rouge">reorder(phy, order = "cladewise")</code></p>

<p>Some of the above functions have the option interactive (FALSE by default) and many are generic.</p>

<p>And some functions to explore the Baumraum (tree space)
<code class="language-plaintext highlighter-rouge">nni</code>, <code class="language-plaintext highlighter-rouge">rNNI</code>, <code class="language-plaintext highlighter-rouge">rSPR</code>, <code class="language-plaintext highlighter-rouge">allTrees</code>.</p>

<h2 id="comparative-methods-and-biogeography">Comparative Methods and Biogeography</h2>

<p>We usually don’t simply stop when we have the tree. We often want to do some downstream analsyses using that phylogeny. This, in fact, is the whole point of the workshop. Broadly, phylogenetic comparative methods typically cluster into discrete and continuous trait methods. Continuous trait methods are for traits that can’t be broken into discrete units. This often includes weights, measurements, and ratios. Discrete traits are, as the name implies, able to be broken down into discrete units. Biogeography often falls into this classification.</p>

<p>Both continuous and discrete trait methods (including biogeography) will typically use a model of evolution to trace a character from the tips of a tree to the root. As an example, let’s a take a quick look at an ancestral state estimation.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">geiger</span><span class="p">)</span><span class="w">
</span><span class="n">q</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">rbind</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">-.5</span><span class="p">,</span><span class="w"> </span><span class="m">.5</span><span class="p">),</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">.5</span><span class="p">,</span><span class="w"> </span><span class="m">-.5</span><span class="p">)))</span><span class="w">

</span><span class="n">trait</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">geiger</span><span class="o">::</span><span class="n">sim.char</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span><span class="w"> </span><span class="n">par</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">q</span><span class="p">,</span><span class="w"> </span><span class="n">model</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"discrete"</span><span class="p">,</span><span class="w"> </span><span class="n">n</span><span class="o">=</span><span class="m">1</span><span class="p">)</span><span class="w">
</span><span class="n">fit</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ace</span><span class="p">(</span><span class="n">trait</span><span class="p">,</span><span class="w"> </span><span class="n">tree</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"discrete"</span><span class="p">,</span><span class="w"> </span><span class="n">model</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ER"</span><span class="p">)</span><span class="w">
</span><span class="n">fit</span><span class="w">
</span></code></pre></div></div>
<p>Let’s walk through this output:</p>

<p><em>Log-likelihood</em> is the likelihood of the data given the Equal-Rates model. The equal rates model can be seen right below - this model says that you are equally likely to go from a state 1 to a state 2 as vice versa - some of you may know this as the Mk model of Lewis (2001) or the Jukes-Cantor model (1969). 
<em>Scaled likelihoods at the root</em> is the likelihood of a state 1 or 2 at the root. We see these are nearly equivalent. Look at the tree - why should this be?</p>

<p>Before we visualize this, we’re going to install one more R Package that we didn’t install prior to the workshop. Many, if not most, R packages are available via CRAN, which is a central packaging repository. However, others are available on GitHub. There are also alternative R package repositories, such as BioConductor. This is where ggtree lives. We will now install it:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="n">require</span><span class="p">(</span><span class="s2">"BiocManager"</span><span class="p">,</span><span class="w"> </span><span class="n">quietly</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))</span><span class="w">
    </span><span class="n">install.packages</span><span class="p">(</span><span class="s2">"BiocManager"</span><span class="p">)</span><span class="w">

</span><span class="n">BiocManager</span><span class="o">::</span><span class="n">install</span><span class="p">(</span><span class="s2">"ggtree"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>With ggtree installed, let’s make some pretty graphics.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">ggtree</span><span class="p">)</span><span class="w">

</span><span class="n">plot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ace</span><span class="p">(</span><span class="n">trait</span><span class="p">,</span><span class="w"> </span><span class="n">tree</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"discrete"</span><span class="p">,</span><span class="w"> </span><span class="n">model</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ER"</span><span class="p">)</span><span class="w">

</span><span class="n">plot</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span><span class="w"> </span><span class="n">cex</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">adj</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0.2</span><span class="p">),</span><span class="w"> </span><span class="n">type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"fan"</span><span class="p">,</span><span class="w"> 
     </span><span class="n">no.margin</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">show.tip.label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span><span class="n">tiplabels</span><span class="p">(</span><span class="n">pch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">16</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="w"> </span><span class="s2">"lightblue"</span><span class="p">,</span><span class="w">
               </span><span class="s2">"chocolate"</span><span class="p">))</span><span class="w">
</span><span class="n">nodelabels</span><span class="p">(</span><span class="n">pie</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">fit</span><span class="o">$</span><span class="n">lik.anc</span><span class="p">,</span><span class="w"> </span><span class="n">cex</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w">
</span><span class="n">legend</span><span class="p">(</span><span class="s2">"bottomleft"</span><span class="p">,</span><span class="w"> </span><span class="n">pch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">15</span><span class="p">,</span><span class="w"> </span><span class="n">bty</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"n"</span><span class="p">,</span><span class="w">
       </span><span class="n">legend</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"0"</span><span class="p">,</span><span class="w"> </span><span class="s2">"1"</span><span class="p">),</span><span class="w"> 
       </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="w"> </span><span class="s2">"lightblue"</span><span class="p">,</span><span class="w">
               </span><span class="s2">"chocolate"</span><span class="p">))</span><span class="w">
               

</span></code></pre></div></div>

<p>Some of the earliest ways of mapping biogeography to trees were based on parsimony. In the time since, it became common to use methods like the above, while treating a landmas as a discrete state, finally, newer models, such as DEC, became more common to use for biogeography. Let’s grab a dataset and take a ramble through some of these methods.</p>

<p>First, we’ll simulate a tree:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="w">
</span><span class="n">big_tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rtree</span><span class="p">(</span><span class="m">45</span><span class="p">)</span><span class="w">
</span><span class="n">big_tree</span><span class="w">

</span></code></pre></div></div>
<p><code class="language-plaintext highlighter-rouge">rtree</code> as a function simulates a random tree. Now let’s add some data to it. This time, we will simulate a tree using a method called “All Rates Different.” Unlike the Mk model or the Jukes-Cantor, this implies that the transition rates are all different between possible character states.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">geiger</span><span class="p">)</span><span class="w">

</span><span class="n">geo</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rTraitDisc</span><span class="p">(</span><span class="n">big_tree</span><span class="p">,</span><span class="w"> </span><span class="s2">"ARD"</span><span class="p">,</span><span class="w"> </span><span class="n">rate</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0.1</span><span class="p">,</span><span class="w"> </span><span class="m">0.2</span><span class="p">))</span><span class="w">
</span><span class="n">fit</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ace</span><span class="p">(</span><span class="n">geo</span><span class="p">,</span><span class="w"> </span><span class="n">big_tree</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"discrete"</span><span class="p">,</span><span class="w"> </span><span class="n">model</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ER"</span><span class="p">)</span><span class="w">
</span><span class="n">fit</span><span class="w">
</span></code></pre></div></div>

<h3 id="exercise">Exercise:</h3>

<ul>
  <li>Try and plot the data from above to a tree.</li>
  <li>What did we do wrong when we fit the trait model above?
    <ul>
      <li>Can you fix it?</li>
      <li>How do you know that you’ve done something better by fixing it? What in this output suggests this?</li>
    </ul>
  </li>
  <li>When you plot an ARD tree, how does it differ from the ER tree?</li>
</ul>

<p>## Explicitly biogeographic models</p>

<p>The above is useful. However, we may want models with parameters that actually carry biogeographic meaning. The most common and well-used of these is the DEC model, standing for Dispersal-Extirpation-Cladogenesis. The idea of this is that a lineage could be dispersing, being removed from an area, or speciating. Described by Ree in 2005, this has gone on to be one of the more popular approaches to understanding phylogeography in deep time. R methods for doing this have fallen apart in the past few years. If there is interest, I’m happy to run this in RevBayes later this week.</p>

<h1 id="references">References</h1>

<p>Jukes T.H., Cantor C.R. 1969. Evolution of Protein Molecules. Mammalian Protein Metabolism. 3:21–132. 10.1016/B978-1-4832-3211-9.50009-7
Lewis P.O. 2001. A Likelihood Approach to Estimating Phylogeny from Discrete Morphological Character Data. Systematic Biology. 50:913–925. 10.1080/106351501753462876</p>

<p>Paradis E. (2012) Definition of Formats for Coding Phylogenetic Trees in R <a href="http://ape-package.ird.fr/misc/FormatTreeR_24Oct2012.pdf">url</a></p>

<p>Paradis E. (2012) Analysis of Phylogenetics and Evolution with R, 2nd ed., Springer, New York</p>

<p>Paradis E. &amp; Schliep K. (2019) ape 5.0: an environment for modern phylogenetics and evolutionary analyses in R, Bioinformatics 35 (3), 526-528</p>

<p>Ree R.H., Moore B.R., Webb C.O., Donoghue M.J., Crandall K. 2005. A likelihood framework for inferring the evolution of geographic range on phylogenetic trees. Evolution. 59:2299–2311. 10.1111/j.0014-3820.2005.tb00940.x</p>

<p>Schliep K.P. 2011. phangorn: phylogenetic analysis in R. Bioinformatics, 27(4) 592-593</p>



<ul class="taxonomy__index">
  
  
    <li>
      <a href="#2023">
        <strong>2023</strong> <span class="taxonomy__count">4</span>
      </a>
    </li>
  
    <li>
      <a href="#2019">
        <strong>2019</strong> <span class="taxonomy__count">1</span>
      </a>
    </li>
  
</ul>




  <section id="2023" class="taxonomy__section">
    <h2 class="archive__subtitle">2023</h2>
    <div class="entries-list">
      
        



<div class="list__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title no_toc" itemprop="headline">
      
        <a href="/updates/02_Biome-evolution/" rel="permalink">Lesson 5 - Biome evolution
</a>
      
    </h2>
    

  <p class="page__meta">
    

    

    
      
      

      <span class="page__meta-readtime">
        <i class="far fa-clock" aria-hidden="true"></i>
        
          1 minute read
        
      </span>
    
  </p>


    <p class="archive__item-excerpt" itemprop="description">
</p>
  </article>
</div>

      
        



<div class="list__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title no_toc" itemprop="headline">
      
        <a href="/updates/02_Biogeographic-data-types/" rel="permalink">Lesson 4 - Biogeographic data types
</a>
      
    </h2>
    

  <p class="page__meta">
    

    

    
      
      

      <span class="page__meta-readtime">
        <i class="far fa-clock" aria-hidden="true"></i>
        
          3 minute read
        
      </span>
    
  </p>


    <p class="archive__item-excerpt" itemprop="description">
</p>
  </article>
</div>

      
        



<div class="list__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title no_toc" itemprop="headline">
      
        <a href="/updates/02_Tree-structure/" rel="permalink">Lesson 2 - Tree structure
</a>
      
    </h2>
    

  <p class="page__meta">
    

    

    
      
      

      <span class="page__meta-readtime">
        <i class="far fa-clock" aria-hidden="true"></i>
        
          10 minute read
        
      </span>
    
  </p>


    <p class="archive__item-excerpt" itemprop="description">
</p>
  </article>
</div>

      
        



<div class="list__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title no_toc" itemprop="headline">
      
        <a href="/updates/01_Packaging-R-Code-Wright/" rel="permalink">Lesson 1 - Packaging R Code
</a>
      
    </h2>
    

  <p class="page__meta">
    

    

    
      
      

      <span class="page__meta-readtime">
        <i class="far fa-clock" aria-hidden="true"></i>
        
          5 minute read
        
      </span>
    
  </p>


    <p class="archive__item-excerpt" itemprop="description">
</p>
  </article>
</div>

      
    </div>
    <a href="#page-title" class="back-to-top">Back to Top &uarr;</a>
  </section>

  <section id="2019" class="taxonomy__section">
    <h2 class="archive__subtitle">2019</h2>
    <div class="entries-list">
      
        



<div class="list__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title no_toc" itemprop="headline">
      
        <a href="/03-RPhylo/" rel="permalink">Tree structure
</a>
      
    </h2>
    

  <p class="page__meta">
    

    

    
      
      

      <span class="page__meta-readtime">
        <i class="far fa-clock" aria-hidden="true"></i>
        
          10 minute read
        
      </span>
    
  </p>


    <p class="archive__item-excerpt" itemprop="description">```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(class.source=’fold-show’)
knitr::opts_chunk$set(collapse = TRUE)
library...</p>
  </article>
</div>

      
    </div>
    <a href="#page-title" class="back-to-top">Back to Top &uarr;</a>
  </section>


  </div>
</div>
    </div>

    

    <div id="footer" class="page__footer">
      <footer>
        <!-- start custom footer snippets -->

<!-- end custom footer snippets -->
        <div class="page__footer-follow">
  <ul class="social-icons">
    
    
      <li><a target="_blank" rel="noopener noreferrer" href="https://twitter.com/Barnabas_Daru"><i class="fab fa-fw fa-twitter-square" aria-hidden="true"></i> Twitter</a></li>
    
    
    
    
    
    <li><a target="_blank" rel="noopener noreferrer" href="http://localhost:4000/feed.xml"><i class="fas fa-fw fa-rss-square" aria-hidden="true"></i> Feed</a></li>
  </ul>
</div>

<div class="page__footer-copyright">
  &copy; 2023 Biogeography and Macroecology 2023. <br>

      </footer>
    </div>

    
  <script src="/assets/js/main.min.js"></script>










  </body>
</html>
