<!doctype html>
<!--
  Minimal Mistakes Jekyll Theme 4.24.0 by Michael Rose
  Copyright 2013-2020 Michael Rose - mademistakes.com | @mmistakes
  Free for personal and commercial use under the MIT license
  https://github.com/mmistakes/minimal-mistakes/blob/master/LICENSE
-->
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">

<!-- begin _includes/seo.html --><title>Lesson 4 - Biogeographic data types - Biogeography and Macroecology 2023</title>
<meta name="description" content="">



<meta property="og:type" content="article">
<meta property="og:locale" content="en_US">
<meta property="og:site_name" content="Biogeography and Macroecology 2023">
<meta property="og:title" content="Lesson 4 - Biogeographic data types">
<meta property="og:url" content="http://localhost:4000/updates/02_Biogeographic-data-types/">




  <meta property="og:image" content="http://localhost:4000/assets/images/tulane.jpg">



  <meta name="twitter:site" content="@Barnabas_Daru">
  <meta name="twitter:title" content="Lesson 4 - Biogeographic data types">
  <meta name="twitter:description" content="">
  <meta name="twitter:url" content="http://localhost:4000/updates/02_Biogeographic-data-types/">

  
    <meta name="twitter:card" content="summary_large_image">
    <meta name="twitter:image" content="http://localhost:4000/assets/images/tulane.jpg">
  

  



  <meta property="article:published_time" content="2023-12-12T00:00:00-06:00">





  

  


<link rel="canonical" href="http://localhost:4000/updates/02_Biogeographic-data-types/">




<script type="application/ld+json">
  {
    "@context": "https://schema.org",
    
      "@type": "Person",
      "name": null,
      "url": "http://localhost:4000/",
      "sameAs": ["https://github.com/darunabas/phyloregion-course-2022","https://twitter.com/Barnabas_Daru"]
    
  }
</script>







<!-- end _includes/seo.html -->



  <link href="/feed.xml" type="application/atom+xml" rel="alternate" title="Biogeography and Macroecology 2023 Feed">


<!-- https://t.co/dKP3o1e -->
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<script>
  document.documentElement.className = document.documentElement.className.replace(/\bno-js\b/g, '') + ' js ';
</script>

<!-- For all browsers -->
<link rel="stylesheet" href="/assets/css/main.css">
<link rel="preload" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5/css/all.min.css" as="style" onload="this.onload=null;this.rel='stylesheet'">
<noscript><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5/css/all.min.css"></noscript>



    <!-- start custom head snippets -->

<!-- insert favicons. use https://realfavicongenerator.net/ -->

<!-- end custom head snippets -->

  </head>

  <body class="layout--posts wide">
    <nav class="skip-links">
  <ul>
    <li><a href="#site-nav" class="screen-reader-shortcut">Skip to primary navigation</a></li>
    <li><a href="#main" class="screen-reader-shortcut">Skip to content</a></li>
    <li><a href="#footer" class="screen-reader-shortcut">Skip to footer</a></li>
  </ul>
</nav>

    <!--[if lt IE 9]>
<div class="notice--danger align-center" style="margin: 0;">You are using an <strong>outdated</strong> browser. Please <a href="https://browsehappy.com/">upgrade your browser</a> to improve your experience.</div>
<![endif]-->

    

<div class="masthead">
  <div class="masthead__inner-wrap">
    <div class="masthead__menu">
      <nav id="site-nav" class="greedy-nav">
        
        <a class="site-title" href="/">
          Biogeography and Macroecology 2023
          
        </a>
        <ul class="visible-links"><li class="masthead__menu-item">
              <a href="/team/">Our Team</a>
            </li><li class="masthead__menu-item">
              <a href="/schedule/">Schedule</a>
            </li><li class="masthead__menu-item">
              <a href="/blog/">Content</a>
            </li><li class="masthead__menu-item">
              <a href="/conduct/">Conduct</a>
            </li><li class="masthead__menu-item">
              <a href="/resources/">Resources</a>
            </li></ul>
        
        <button class="greedy-nav__toggle hidden" type="button">
          <span class="visually-hidden">Toggle menu</span>
          <div class="navicon"></div>
        </button>
        <ul class="hidden-links hidden"></ul>
      </nav>
    </div>
  </div>
</div>


    <div class="initial-content">
      
  







<div class="page__hero--overlay"
  style=" background-image: linear-gradient(rgba(0, 0, 0, 0.2), rgba(0, 0, 0, 0.2)), url('/assets/images/tulane.jpg');"
>
  
    <div class="wrapper">
      <h1 id="page-title" class="page__title" itemprop="headline">
        
          Lesson 4 - Biogeographic data types

        
      </h1>
      
        <p class="page__lead">
</p>
      
      

  <p class="page__meta">
    

    

    
      
      

      <span class="page__meta-readtime">
        <i class="far fa-clock" aria-hidden="true"></i>
        
          9 minute read
        
      </span>
    
  </p>


      
      
    </div>
  
  
</div>





<div id="main" role="main">
  


  <div class="archive">
    
    <!-- Look the author details up from the site config. -->


<!-- Output author details if some exist. -->


<h2 id="lecture-4">Lecture 4</h2>
<h2 id="methods-to-infer-biogeography-patternsprocesses">Methods to infer biogeography patterns/processes</h2>

<h3 id="fossils-eg-pollen">Fossils (e.g., pollen)</h3>
<ul>
  <li>Paleoenvironments</li>
  <li>Pollen records</li>
  <li>Dating</li>
</ul>

<h3 id="dnamolecular">DNA/molecular</h3>
<ul>
  <li>Phylogenetic trees</li>
  <li>Ancestral area reconstruction</li>
  <li>Extinction</li>
  <li>Speciation</li>
  <li>Dispersal</li>
</ul>

<h3 id="species-occurrences">Species occurrences</h3>
<ul>
  <li>GBIF</li>
  <li>Point records</li>
  <li>Raster layers</li>
  <li>Range polygons</li>
</ul>

<h2 id="fossil-data">Fossil data</h2>
<p>Biogeographers study fossils to learn about past geological events, climates, and evolution.  Fossils are preserved remains or traces of living things. Fossils normally form in sedimentary rock. Hard parts are the only parts of an organism that leaves a fossil.
Examples: Bones, shells, teeth, seeds, and woody stems.</p>

<p>However, fossils are not available for most taxonomic groups e.g., soft-body organisms, and geographic regions e.g., the tropics.</p>

<h2 id="dna-sequence-data">DNA sequence data</h2>
<p>Perhaps the most important method for inferring phylogenetic relationships of life is that of acquiring DNA sequences. 
DNA sequence data basically refers to the sequence of nucleotides (adenine = A, cytosine = C, guanine = 0, or thymine = T) in a particular region of the DNA of a given taxon.
DNA sequences are aligned and used for phylogenetic reconstruction.</p>

<h2 id="other-types-of-data-for-biogeographic-research">Other types of data for biogeographic research</h2>
<p>Biogeographical data can be linked to a wide range of organismic (e.g., taxonomic, functional, phylogenetic) and environmental (e.g., climate, soil, topography) information.</p>

<h2 id="species-occurrence-data">Species occurrence data</h2>
<p>Species occurrences can be represented by point records, vegetation plots, checklists, range polygon maps, or raster layers.</p>

<h3 id="1-point-records">1. Point records</h3>
<p>Point occurrences indicate points on a map where a species has been recorded. Point records often show localities where museum or herbarium specimens have been collected or observed. They are often prevalent with sampling gaps and biases.</p>

<p>For this lesson, we will download some geographic data from the Global Biodiversity Information Facility (GBIF, <a href="https://www.gbif.org/">https://www.gbif.org/</a>). You can optionally use the <code class="language-plaintext highlighter-rouge">geodata</code> R package to download geographic data. The <code class="language-plaintext highlighter-rouge">geodata</code> package has functions for downloading different types of geographic data including climate, crops, elevation, land use, soil, species occurrence, accessibility, administrative boundaries and other data. This means that internet access is required to download the datasets.</p>

<p>Here, we will directly download occurrence records from GBIF for all the oak species in the genus <em>Quercus</em> (Plantae » Tracheophyta » Magnoliopsida » Fagales » Fagaceae » Quercus, n = 5,465,327 records) <a href="https://doi.org/10.15468/dl.3b7amt">https://doi.org/10.15468/dl.3b7amt</a>. Note, the compressed data size is: 465.7 MB and the actual data size is 3.5GB. So be sure to have some space on your computer to download it. Next, we will process this dataset in several steps:</p>

<h4 id="step-1-data-cleaning">Step 1: Data cleaning</h4>
<p>We will use the R package <code class="language-plaintext highlighter-rouge">CoordinateCleaner</code> to clean the records to remove duplicates and records with erroneous localizations. Depending on the goals of your study, there are many ways of cleaning the dataset. For example, you may want to remove records with questionable scientific names by querying against currently accepted taxonomies for the taxonomic group of interest.</p>

<p>Since the genus <em>Quercus</em> is widely distributed in the Northern Hemisphere, we will clean it by removing all records south of the equator, i.e., latitude &lt; 0</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">u</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"/Users/datasets/"</span><span class="w">
</span><span class="n">d</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.table</span><span class="o">::</span><span class="n">fread</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="n">u</span><span class="p">,</span><span class="w"> </span><span class="s2">"0026030-231120084113126.csv"</span><span class="p">))</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">d</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>Inspecting the data, you will notice several fields, but since we’re interested in geography, we’ll only keep the species and geographic coordinates (longitude and latitude fields). Note: most geographic analyses in R requires the coordinates arranged as longitude before latitude.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">q</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">d</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"species"</span><span class="p">,</span><span class="w"> </span><span class="s2">"decimalLongitude"</span><span class="p">,</span><span class="w"> </span><span class="s2">"decimalLatitude"</span><span class="p">)]</span><span class="w"> 
</span><span class="nf">names</span><span class="p">(</span><span class="n">q</span><span class="p">)[</span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">3</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"lon"</span><span class="p">,</span><span class="w"> </span><span class="s2">"lat"</span><span class="p">)</span><span class="w">
</span><span class="n">q</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">q</span><span class="p">[</span><span class="o">!</span><span class="p">(</span><span class="n">q</span><span class="o">$</span><span class="n">species</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="s2">""</span><span class="p">),]</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="n">unique</span><span class="p">()</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="n">na.omit</span><span class="p">()</span><span class="w">
</span><span class="nf">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">q</span><span class="o">$</span><span class="n">species</span><span class="p">))</span><span class="w"> </span><span class="c1"># count the number of species = 699</span><span class="w">
</span><span class="n">r</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">q</span><span class="p">[</span><span class="n">q</span><span class="o">$</span><span class="n">lat</span><span class="o">&gt;</span><span class="m">0</span><span class="p">,]</span><span class="w">
</span></code></pre></div></div>

<p>We can plot the distribution for one of the species, White Oak (<em>Quercus alba</em>). Note: don’t attempt to plot the entire dataset as it can freeze your computer.’</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ex</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">r</span><span class="p">[</span><span class="n">r</span><span class="o">$</span><span class="n">species</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="s2">"Quercus alba"</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="n">ex</span><span class="o">$</span><span class="n">lon</span><span class="p">,</span><span class="w"> </span><span class="n">ex</span><span class="o">$</span><span class="n">lat</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>
<p>Depending on the data source, an ultimate goal is to convert the distribution data to a community data frame across different sites and geographic extents for downstream analyses.</p>

<p>Thus, it is important to define the study extent. For illustration purposes, we will download vector polygons for the lower 48 Contiguous United States (not including AK, HI and PR) from
<a href="https://hub.arcgis.com/datasets/1b02c87f62d24508970dc1a6df80c98e/explore">https://hub.arcgis.com/datasets/1b02c87f62d24508970dc1a6df80c98e/explore</a>. Simply download and store the folder on your computer locally. You can also use the function <code class="language-plaintext highlighter-rouge">world</code> in the package <code class="language-plaintext highlighter-rouge">geodata</code> to get the borders for all the countries in the world.</p>

<p>First, we will use the R package <code class="language-plaintext highlighter-rouge">terra</code> for most of our geographical manipulations.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">terra</span><span class="p">)</span><span class="w">
</span><span class="n">v</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">vect</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="n">u</span><span class="p">,</span><span class="w"> </span><span class="s2">"States_shapefile-shp/States_shapefile.shp"</span><span class="p">))</span><span class="w">
</span><span class="n">v</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">v</span><span class="p">[</span><span class="o">!</span><span class="p">(</span><span class="n">v</span><span class="o">$</span><span class="n">State_Name</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"ALASKA"</span><span class="p">,</span><span class="w"> </span><span class="s2">"HAWAII"</span><span class="p">)),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span></code></pre></div></div>
<p>Let’s assume that we want to define equal size quadrats or grid cells or fishnet of 1 degree resolution across the entire landscape of conterminous USA.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">s</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rast</span><span class="p">(</span><span class="n">res</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">ext</span><span class="p">(</span><span class="n">v</span><span class="p">))</span><span class="w">
</span><span class="n">crs</span><span class="p">(</span><span class="n">s</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"epsg:4326"</span><span class="w">
</span><span class="n">m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.polygons</span><span class="p">(</span><span class="n">s</span><span class="p">,</span><span class="w"> </span><span class="n">dissolve</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span><span class="n">m</span><span class="o">$</span><span class="n">grids</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste0</span><span class="p">(</span><span class="s2">"v"</span><span class="p">,</span><span class="w"> </span><span class="nf">seq_len</span><span class="p">(</span><span class="n">nrow</span><span class="p">(</span><span class="n">m</span><span class="p">)))</span><span class="w">
</span><span class="n">m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">m</span><span class="p">[,</span><span class="w"> </span><span class="s2">"grids"</span><span class="p">]</span><span class="w">
</span><span class="n">z</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">m</span><span class="p">[</span><span class="n">v</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
</span></code></pre></div></div>
<p>We can write a function to automate this process as follows:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">myfishnet</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">mask</span><span class="p">,</span><span class="w"> </span><span class="n">res</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
  </span><span class="n">s</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rast</span><span class="p">(</span><span class="n">res</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">res</span><span class="p">,</span><span class="w"> </span><span class="n">ext</span><span class="p">(</span><span class="n">mask</span><span class="p">))</span><span class="w">
  </span><span class="n">crs</span><span class="p">(</span><span class="n">s</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"epsg:4326"</span><span class="w">
  </span><span class="n">m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.polygons</span><span class="p">(</span><span class="n">s</span><span class="p">,</span><span class="w"> </span><span class="n">dissolve</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
  </span><span class="n">m</span><span class="o">$</span><span class="n">grids</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste0</span><span class="p">(</span><span class="s2">"v"</span><span class="p">,</span><span class="w"> </span><span class="nf">seq_len</span><span class="p">(</span><span class="n">nrow</span><span class="p">(</span><span class="n">m</span><span class="p">)))</span><span class="w">
  </span><span class="n">m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">m</span><span class="p">[,</span><span class="w"> </span><span class="s2">"grids"</span><span class="p">]</span><span class="w">
  </span><span class="n">m</span><span class="p">[</span><span class="n">mask</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="n">my_grids</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">myfishnet</span><span class="p">(</span><span class="n">mask</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">v</span><span class="p">,</span><span class="w"> </span><span class="n">res</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h4 id="step-2-convert-raw-input-distribution-data-to-community">Step 2: Convert raw input distribution data to community</h4>
<p>Here, we will convert the raw point occurrence data to a community composition data frame across the quadrats we generated for mainland USA.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">b</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">vect</span><span class="p">(</span><span class="n">pt</span><span class="p">)</span><span class="w">
</span><span class="n">j</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">relate</span><span class="p">(</span><span class="n">b</span><span class="p">,</span><span class="w"> </span><span class="n">my_grids</span><span class="p">,</span><span class="w"> </span><span class="s2">"intersects"</span><span class="p">,</span><span class="w"> </span><span class="n">pairs</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">my_grids</span><span class="o">$</span><span class="n">grids</span><span class="p">[</span><span class="n">j</span><span class="p">[,</span><span class="w"> </span><span class="m">2</span><span class="p">]]</span><span class="w">
</span><span class="n">spp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">b</span><span class="o">$</span><span class="n">species</span><span class="p">[</span><span class="n">j</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">]]</span><span class="w">
</span><span class="n">r</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">species</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">spp</span><span class="p">,</span><span class="w"> </span><span class="n">grids</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">y</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>
<p>We can easily convert this long-form dataset to a community composition matrix of 1s and 0s, with species as columns and rows as spatial cells or communities. However, such data storage can be computationally challenging and expensive. To overcome this challenge, we will use a sparse matrix approach using the R package <code class="language-plaintext highlighter-rouge">Matrix</code>. Because a sparse matrix is comprised mostly of 0s, it only stores the non-zero entries, from which several measures of biodiversity can be calculated.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">Matrix</span><span class="p">)</span><span class="w">
</span><span class="n">x</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">r</span><span class="p">)</span><span class="w">
</span><span class="n">x</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">x</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"grids"</span><span class="p">,</span><span class="w"> </span><span class="s2">"species"</span><span class="p">)]</span><span class="w">
</span><span class="nf">names</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"grids"</span><span class="p">,</span><span class="w"> </span><span class="s2">"species"</span><span class="p">)</span><span class="w">
</span><span class="n">grids</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">x</span><span class="o">$</span><span class="n">grids</span><span class="p">))</span><span class="w">
</span><span class="n">species</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">x</span><span class="o">$</span><span class="n">species</span><span class="p">))</span><span class="w">
</span><span class="n">res</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sparseMatrix</span><span class="p">(</span><span class="nf">as.integer</span><span class="p">(</span><span class="n">grids</span><span class="p">),</span><span class="w"> </span><span class="nf">as.integer</span><span class="p">(</span><span class="n">species</span><span class="p">),</span><span class="w"> 
                    </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">dimnames</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">levels</span><span class="p">(</span><span class="n">grids</span><span class="p">),</span><span class="w"> </span><span class="n">levels</span><span class="p">(</span><span class="n">species</span><span class="p">)))</span><span class="w">
</span></code></pre></div></div>

<p>This can also be written up as a function and we can call it <code class="language-plaintext highlighter-rouge">mylong2sparse</code></p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">mylong2sparse</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">grids</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grids"</span><span class="p">,</span><span class="w"> </span><span class="n">species</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"species"</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
  </span><span class="n">x</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w">
  </span><span class="n">x</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">x</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">grids</span><span class="p">,</span><span class="w"> </span><span class="n">species</span><span class="p">)]</span><span class="w">
  </span><span class="nf">names</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"grids"</span><span class="p">,</span><span class="w"> </span><span class="s2">"species"</span><span class="p">)</span><span class="w">
  </span><span class="n">grids</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">x</span><span class="o">$</span><span class="n">grids</span><span class="p">))</span><span class="w">
  </span><span class="n">species</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">x</span><span class="o">$</span><span class="n">species</span><span class="p">))</span><span class="w">
  </span><span class="n">res</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sparseMatrix</span><span class="p">(</span><span class="nf">as.integer</span><span class="p">(</span><span class="n">grids</span><span class="p">),</span><span class="w"> </span><span class="nf">as.integer</span><span class="p">(</span><span class="n">species</span><span class="p">),</span><span class="w"> 
                      </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">dimnames</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">levels</span><span class="p">(</span><span class="n">grids</span><span class="p">),</span><span class="w"> </span><span class="n">levels</span><span class="p">(</span><span class="n">species</span><span class="p">)))</span><span class="w">
  </span><span class="n">res</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="n">test1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mylong2sparse</span><span class="p">(</span><span class="n">r</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>We can wrap this and the previous chunks of code into a function called <code class="language-plaintext highlighter-rouge">mypoints2comm</code></p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">mypoints2comm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span><span class="w"> </span><span class="n">res</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.25</span><span class="p">,</span><span class="w"> </span><span class="n">pol.grids</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">...</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
  </span><span class="n">dat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">na.omit</span><span class="p">(</span><span class="n">dat</span><span class="p">)</span><span class="w">
  </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">pol.grids</span><span class="p">))</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">pol.grids</span><span class="p">[,</span><span class="w"> </span><span class="n">grep</span><span class="p">(</span><span class="s2">"grids"</span><span class="p">,</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">pol.grids</span><span class="p">)),</span><span class="w"> </span><span class="n">drop</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">]</span><span class="w">
  </span><span class="p">}</span><span class="w">
  </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">fishnet</span><span class="p">(</span><span class="n">mask</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ext</span><span class="p">(</span><span class="n">vect</span><span class="p">(</span><span class="n">dat</span><span class="p">)),</span><span class="w"> </span><span class="n">res</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">res</span><span class="p">)</span><span class="w">
    </span><span class="n">crs</span><span class="p">(</span><span class="n">m</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"+proj=longlat +datum=WGS84"</span><span class="w">
  </span><span class="p">}</span><span class="w">
  </span><span class="n">pj</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"+proj=longlat +datum=WGS84"</span><span class="w">
  </span><span class="n">m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">suppressWarnings</span><span class="p">(</span><span class="nf">invisible</span><span class="p">(</span><span class="n">project</span><span class="p">(</span><span class="n">m</span><span class="p">,</span><span class="w"> </span><span class="s2">"epsg:4326"</span><span class="p">)))</span><span class="w">
  </span><span class="n">crs</span><span class="p">(</span><span class="n">m</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">pj</span><span class="w">
  </span><span class="n">b</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">vect</span><span class="p">(</span><span class="n">dat</span><span class="p">)</span><span class="w">
  </span><span class="n">j</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">relate</span><span class="p">(</span><span class="n">b</span><span class="p">,</span><span class="w"> </span><span class="n">m</span><span class="p">,</span><span class="w"> </span><span class="s2">"intersects"</span><span class="p">,</span><span class="w"> </span><span class="n">pairs</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
  </span><span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">m</span><span class="o">$</span><span class="n">grids</span><span class="p">[</span><span class="n">j</span><span class="p">[,</span><span class="w"> </span><span class="m">2</span><span class="p">]]</span><span class="w">
  </span><span class="n">spp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">b</span><span class="o">$</span><span class="n">species</span><span class="p">[</span><span class="n">j</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">]]</span><span class="w">
  </span><span class="n">r</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">species</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">spp</span><span class="p">,</span><span class="w"> </span><span class="n">grids</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">y</span><span class="p">)</span><span class="w">
  </span><span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mylong2sparse</span><span class="p">(</span><span class="n">r</span><span class="p">)</span><span class="w">
  </span><span class="n">tmp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">grids</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">row.names</span><span class="p">(</span><span class="n">y</span><span class="p">),</span><span class="w"> </span><span class="n">abundance</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Matrix</span><span class="o">::</span><span class="n">rowSums</span><span class="p">(</span><span class="n">y</span><span class="p">),</span><span class="w"> 
                    </span><span class="n">richness</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Matrix</span><span class="o">::</span><span class="n">rowSums</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0</span><span class="p">))</span><span class="w">
  </span><span class="n">z</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">m</span><span class="p">,</span><span class="w"> </span><span class="n">tmp</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grids"</span><span class="p">)</span><span class="w">
  </span><span class="nf">return</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">comm_dat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">y</span><span class="p">,</span><span class="w"> </span><span class="n">map</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">z</span><span class="p">))</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="n">z</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mypoints2comm</span><span class="p">(</span><span class="n">dat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pt</span><span class="p">,</span><span class="w"> </span><span class="n">pol.grids</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">my_grids</span><span class="p">)</span><span class="w">
</span><span class="nf">names</span><span class="p">(</span><span class="n">z</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>
<p>The output is a list of two objects: <code class="language-plaintext highlighter-rouge">comm_dat</code>: (sparse) community matrix and <code class="language-plaintext highlighter-rouge">map</code>: vector or raster of grid cells with the values per cell for mapping.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">mp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">z</span><span class="o">$</span><span class="n">map</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">mp</span><span class="p">)</span><span class="w">

</span><span class="n">par</span><span class="p">(</span><span class="n">mfrow</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">1</span><span class="p">))</span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="n">mp</span><span class="p">,</span><span class="w"> </span><span class="s2">"richness"</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"continuous"</span><span class="p">)</span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="n">mp</span><span class="p">,</span><span class="w"> </span><span class="s2">"abundance"</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"continuous"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h3 id="2-range-polygons">2. Range polygons</h3>
<p>Range polygons or range maps show a range as an area, typically shaded, within a boundary.  The boundary line defines the limits of the known distribution of the species. Often much guesswork involved.
Polygons can be derived from the International Union for the Conservation of Nature’s spatial database (https://www.iucnredlist.org/resources/spatial-data-download), published monographs or field guides that have been validated by taxonomic experts.</p>

<p>Next, we will convert the point occurrences to range polygons using the R package <code class="language-plaintext highlighter-rouge">rangeBuilder</code>. This will allow us to generate alpha hulls. We will then crop the alpha hulls to terrestrial areas by removing areas falling in the ocean.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">rangeBuilder</span><span class="p">)</span><span class="w">
</span><span class="nf">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">pt</span><span class="o">$</span><span class="n">species</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>
<p>Because our point occurrence data has 698 species, we should split the data into a manageable size.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">f</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">split</span><span class="p">(</span><span class="n">pt</span><span class="p">,</span><span class="w"> </span><span class="n">pt</span><span class="o">$</span><span class="n">species</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>
<p>Next, we will use the <code class="language-plaintext highlighter-rouge">getDynamicAlphaHull</code> function in the R package <code class="language-plaintext highlighter-rouge">rangeBuilder</code> to generate the alpha hulls. However, the process of generating alphahulls is quite complicated and requires some tweaking of the parameters especially because species occurrence records vary in terms of their sampling – some species are more or less sampled than others. For species that are oversampled, it’s important to thin them to speed up computational time. By contrast, species with very few records should be ‘helped’ by buffering and random sampling around such points. For the spatial thinning, we will first use the <code class="language-plaintext highlighter-rouge">geodata</code> package to download a elevation raster layer for the thinning. Any raster layer would work. We then use the <code class="language-plaintext highlighter-rouge">gridSample</code> function from <code class="language-plaintext highlighter-rouge">dismo</code> for the thinning.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ras</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">geodata</span><span class="o">::</span><span class="n">elevation_global</span><span class="p">(</span><span class="n">res</span><span class="o">=</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">path</span><span class="o">=</span><span class="s2">"/Users/bdaru/Downloads"</span><span class="p">)</span><span class="w">

</span><span class="n">l</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">lapply</span><span class="p">(</span><span class="n">f</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">],</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
  </span><span class="n">sp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">unique</span><span class="p">(</span><span class="n">x</span><span class="o">$</span><span class="n">species</span><span class="p">)</span><span class="w">
  </span><span class="k">if</span><span class="p">(</span><span class="n">nrow</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">5</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">buffer</span><span class="p">(</span><span class="n">vect</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">crs</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"+proj=longlat"</span><span class="p">),</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">50000</span><span class="p">)</span><span class="w">
    </span><span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">spatSample</span><span class="p">(</span><span class="n">y</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">100</span><span class="p">)</span><span class="w">
    </span><span class="n">q</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">geom</span><span class="p">(</span><span class="n">y</span><span class="p">,</span><span class="w"> </span><span class="n">df</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"x"</span><span class="p">,</span><span class="w"> </span><span class="s2">"y"</span><span class="p">)]</span><span class="w">
    </span><span class="nf">names</span><span class="p">(</span><span class="n">q</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"lon"</span><span class="p">,</span><span class="w"> </span><span class="s2">"lat"</span><span class="p">)</span><span class="w">
    </span><span class="n">q</span><span class="o">$</span><span class="n">species</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sp</span><span class="w">
    </span><span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">q</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"species"</span><span class="p">,</span><span class="w"> </span><span class="s2">"lon"</span><span class="p">,</span><span class="w"> </span><span class="s2">"lat"</span><span class="p">)]</span><span class="w">
  </span><span class="p">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="k">if</span><span class="p">(</span><span class="n">nrow</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">2000</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">dismo</span><span class="o">::</span><span class="n">gridSample</span><span class="p">(</span><span class="n">x</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"lon"</span><span class="p">,</span><span class="w"> </span><span class="s2">"lat"</span><span class="p">)],</span><span class="w"> </span><span class="n">r</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ras</span><span class="p">)</span><span class="w">
    </span><span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">species</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sp</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="p">)</span><span class="w">
  </span><span class="p">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">x</span><span class="w">
  </span><span class="p">}</span><span class="w">
  </span><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">tryCatch</span><span class="p">(</span><span class="n">getDynamicAlphaHull</span><span class="p">(</span><span class="n">y</span><span class="p">,</span><span class="w"> </span><span class="n">coordHeaders</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"lon"</span><span class="p">,</span><span class="w"> </span><span class="s2">"lat"</span><span class="p">),</span><span class="w">
                                    </span><span class="n">clipToCoast</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'no'</span><span class="p">,</span><span class="w"> </span><span class="n">initialAlpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w">
                </span><span class="n">error</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="w"> </span><span class="kc">NULL</span><span class="p">)</span><span class="w">
  </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">p</span><span class="p">))</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">q</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">vect</span><span class="p">(</span><span class="n">p</span><span class="p">[[</span><span class="m">1</span><span class="p">]])</span><span class="w">
    </span><span class="n">z</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">crop</span><span class="p">(</span><span class="n">q</span><span class="p">,</span><span class="w"> </span><span class="n">v</span><span class="p">)</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="n">terra</span><span class="o">::</span><span class="n">aggregate</span><span class="p">(</span><span class="n">dissolve</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
    </span><span class="k">if</span><span class="p">(</span><span class="n">geomtype</span><span class="p">(</span><span class="n">z</span><span class="p">)</span><span class="o">==</span><span class="s2">"polygons"</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
      </span><span class="n">z</span><span class="o">$</span><span class="n">binomial</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sp</span><span class="w">
      </span><span class="nf">return</span><span class="p">(</span><span class="n">z</span><span class="p">)</span><span class="w">
    </span><span class="p">}</span><span class="w">
  </span><span class="p">}</span><span class="w"> 
</span><span class="p">})</span><span class="w">
</span><span class="n">out</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">Filter</span><span class="p">(</span><span class="n">Negate</span><span class="p">(</span><span class="n">is.null</span><span class="p">),</span><span class="w"> </span><span class="n">l</span><span class="p">)</span><span class="w">

</span><span class="n">v2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">vect</span><span class="p">(</span><span class="n">out</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>
<p>Just like we did for point records, we can convert the polygon range maps to a community matrix by overlaying them onto quadrats to extract presences and absences within the quadrats. This can be achieved using the <code class="language-plaintext highlighter-rouge">relate</code> function in <code class="language-plaintext highlighter-rouge">terra</code>.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">j</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">relate</span><span class="p">(</span><span class="n">v2</span><span class="p">,</span><span class="w"> </span><span class="n">my_grids</span><span class="p">,</span><span class="w"> </span><span class="s2">"intersects"</span><span class="p">,</span><span class="w"> </span><span class="n">pairs</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">my_grids</span><span class="o">$</span><span class="n">grids</span><span class="p">[</span><span class="n">j</span><span class="p">[,</span><span class="w"> </span><span class="m">2</span><span class="p">]]</span><span class="w">
</span><span class="n">spp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">v2</span><span class="o">$</span><span class="n">binomial</span><span class="p">[</span><span class="n">j</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">]]</span><span class="w">
</span><span class="n">r</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">species</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">spp</span><span class="p">,</span><span class="w"> </span><span class="n">grids</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">y</span><span class="p">)</span><span class="w">

</span><span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mylong2sparse</span><span class="p">(</span><span class="n">r</span><span class="p">)</span><span class="w">
</span><span class="n">tmp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">grids</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">row.names</span><span class="p">(</span><span class="n">y</span><span class="p">),</span><span class="w"> </span><span class="n">abundance</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rowSums</span><span class="p">(</span><span class="n">y</span><span class="p">),</span><span class="w"> 
                  </span><span class="n">richness</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rowSums</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0</span><span class="p">))</span><span class="w">
</span><span class="n">z</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">m</span><span class="p">,</span><span class="w"> </span><span class="n">tmp</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grids"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>We can write it up as a function as follows:</p>
<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">mypolys2comm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span><span class="w"> </span><span class="n">res</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.25</span><span class="p">,</span><span class="w"> </span><span class="n">pol.grids</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">...</span><span class="p">){</span><span class="w">
  </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">pol.grids</span><span class="p">))</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">pol.grids</span><span class="p">[,</span><span class="w"> </span><span class="n">grep</span><span class="p">(</span><span class="s2">"grids"</span><span class="p">,</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">pol.grids</span><span class="p">)),</span><span class="w"> </span><span class="n">drop</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">]</span><span class="w">
  </span><span class="p">}</span><span class="w">
  </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">fishnet</span><span class="p">(</span><span class="n">mask</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ext</span><span class="p">(</span><span class="n">dat</span><span class="p">),</span><span class="w"> </span><span class="n">res</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">res</span><span class="p">)</span><span class="w">
    </span><span class="n">crs</span><span class="p">(</span><span class="n">m</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"+proj=longlat +datum=WGS84"</span><span class="w">
  </span><span class="p">}</span><span class="w">
  </span><span class="n">pj</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"+proj=longlat +datum=WGS84"</span><span class="w">
  </span><span class="n">m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">suppressWarnings</span><span class="p">(</span><span class="nf">invisible</span><span class="p">(</span><span class="n">project</span><span class="p">(</span><span class="n">m</span><span class="p">,</span><span class="w"> </span><span class="s2">"epsg:4326"</span><span class="p">)))</span><span class="w">
  </span><span class="n">crs</span><span class="p">(</span><span class="n">m</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">pj</span><span class="w">
  </span><span class="n">j</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">relate</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span><span class="w"> </span><span class="n">m</span><span class="p">,</span><span class="w"> </span><span class="s2">"intersects"</span><span class="p">,</span><span class="w"> </span><span class="n">pairs</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
  </span><span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">m</span><span class="o">$</span><span class="n">grids</span><span class="p">[</span><span class="n">j</span><span class="p">[,</span><span class="w"> </span><span class="m">2</span><span class="p">]]</span><span class="w">
  </span><span class="n">spp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">dat</span><span class="o">$</span><span class="n">binomial</span><span class="p">[</span><span class="n">j</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">]]</span><span class="w">
  </span><span class="n">r</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">species</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">spp</span><span class="p">,</span><span class="w"> </span><span class="n">grids</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">y</span><span class="p">)</span><span class="w">
  </span><span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">long2sparse</span><span class="p">(</span><span class="n">r</span><span class="p">)</span><span class="w">
  </span><span class="n">tmp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">grids</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">row.names</span><span class="p">(</span><span class="n">y</span><span class="p">),</span><span class="w"> </span><span class="n">abundance</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rowSums</span><span class="p">(</span><span class="n">y</span><span class="p">),</span><span class="w"> 
                    </span><span class="n">richness</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rowSums</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0</span><span class="p">))</span><span class="w">
  </span><span class="n">z</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">m</span><span class="p">,</span><span class="w"> </span><span class="n">tmp</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grids"</span><span class="p">)</span><span class="w">
  </span><span class="nf">return</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">comm_dat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">y</span><span class="p">,</span><span class="w"> </span><span class="n">map</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">z</span><span class="p">))</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="n">test3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mypolys2comm</span><span class="p">(</span><span class="n">dat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">v2</span><span class="p">,</span><span class="w"> </span><span class="n">pol.grids</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">my_grids</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>Similar to <code class="language-plaintext highlighter-rouge">mypoints2comm</code>, the output from <code class="language-plaintext highlighter-rouge">mypolys2comm</code> is also a list of two objects: <code class="language-plaintext highlighter-rouge">comm_dat</code>: (sparse) community matrix and <code class="language-plaintext highlighter-rouge">map</code>: vector or raster of grid cells with the values per cell for mapping.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">mp2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">test3</span><span class="o">$</span><span class="n">map</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">mp2</span><span class="p">)</span><span class="w">

</span><span class="n">par</span><span class="p">(</span><span class="n">mfrow</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">1</span><span class="p">))</span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="n">mp2</span><span class="p">,</span><span class="w"> </span><span class="s2">"richness"</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"continuous"</span><span class="p">)</span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="n">mp2</span><span class="p">,</span><span class="w"> </span><span class="s2">"abundance"</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"continuous"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>
<h4 id="exercise">EXERCISE</h4>
<p>Based on all the tools you have learned so far, download occurrence records for the Eucalyptus species of Australia from GBIF. Read them in, clean, and map species richness and abundances based on point records for the first 100 species across equal area grid cells of 0.5 degrees. Generate range polygons for the first 100 species and use that to generate species richness and abundance of the species. Compare your richness maps from point records against that of polygons using a simple linear regression.</p>

<h4 id="challenge">CHALLENGE</h4>
<p>Write a function to match the species list from point records versus that of polygons.</p>

<h3 id="3-raster-layers">3. Raster layers</h3>
<p>Raster layers maps are generally derived from species distribution modelling. This is an entirely broad subject and will be covered separately as its own topic.</p>



<ul class="taxonomy__index">
  
  
    <li>
      <a href="#2023">
        <strong>2023</strong> <span class="taxonomy__count">4</span>
      </a>
    </li>
  
    <li>
      <a href="#2019">
        <strong>2019</strong> <span class="taxonomy__count">1</span>
      </a>
    </li>
  
</ul>




  <section id="2023" class="taxonomy__section">
    <h2 class="archive__subtitle">2023</h2>
    <div class="entries-list">
      
        



<div class="list__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title no_toc" itemprop="headline">
      
        <a href="/updates/02_Biome-evolution/" rel="permalink">Lesson 5 - Biome evolution
</a>
      
    </h2>
    

  <p class="page__meta">
    

    

    
      
      

      <span class="page__meta-readtime">
        <i class="far fa-clock" aria-hidden="true"></i>
        
          1 minute read
        
      </span>
    
  </p>


    <p class="archive__item-excerpt" itemprop="description">
</p>
  </article>
</div>

      
        



<div class="list__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title no_toc" itemprop="headline">
      
        <a href="/updates/02_Biogeographic-data-types/" rel="permalink">Lesson 4 - Biogeographic data types
</a>
      
    </h2>
    

  <p class="page__meta">
    

    

    
      
      

      <span class="page__meta-readtime">
        <i class="far fa-clock" aria-hidden="true"></i>
        
          9 minute read
        
      </span>
    
  </p>


    <p class="archive__item-excerpt" itemprop="description">
</p>
  </article>
</div>

      
        



<div class="list__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title no_toc" itemprop="headline">
      
        <a href="/updates/02_Tree-structure/" rel="permalink">Lesson 2 - Tree structure
</a>
      
    </h2>
    

  <p class="page__meta">
    

    

    
      
      

      <span class="page__meta-readtime">
        <i class="far fa-clock" aria-hidden="true"></i>
        
          10 minute read
        
      </span>
    
  </p>


    <p class="archive__item-excerpt" itemprop="description">
</p>
  </article>
</div>

      
        



<div class="list__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title no_toc" itemprop="headline">
      
        <a href="/updates/01_Packaging-R-Code-Wright/" rel="permalink">Lesson 1 - Packaging R Code
</a>
      
    </h2>
    

  <p class="page__meta">
    

    

    
      
      

      <span class="page__meta-readtime">
        <i class="far fa-clock" aria-hidden="true"></i>
        
          5 minute read
        
      </span>
    
  </p>


    <p class="archive__item-excerpt" itemprop="description">
</p>
  </article>
</div>

      
    </div>
    <a href="#page-title" class="back-to-top">Back to Top &uarr;</a>
  </section>

  <section id="2019" class="taxonomy__section">
    <h2 class="archive__subtitle">2019</h2>
    <div class="entries-list">
      
        



<div class="list__item">
  <article class="archive__item" itemscope itemtype="https://schema.org/CreativeWork">
    
    <h2 class="archive__item-title no_toc" itemprop="headline">
      
        <a href="/03-RPhylo/" rel="permalink">Tree structure
</a>
      
    </h2>
    

  <p class="page__meta">
    

    

    
      
      

      <span class="page__meta-readtime">
        <i class="far fa-clock" aria-hidden="true"></i>
        
          10 minute read
        
      </span>
    
  </p>


    <p class="archive__item-excerpt" itemprop="description">```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(class.source=’fold-show’)
knitr::opts_chunk$set(collapse = TRUE)
library...</p>
  </article>
</div>

      
    </div>
    <a href="#page-title" class="back-to-top">Back to Top &uarr;</a>
  </section>


  </div>
</div>
    </div>

    

    <div id="footer" class="page__footer">
      <footer>
        <!-- start custom footer snippets -->

<!-- end custom footer snippets -->
        <div class="page__footer-follow">
  <ul class="social-icons">
    
    
      <li><a target="_blank" rel="noopener noreferrer" href="https://twitter.com/Barnabas_Daru"><i class="fab fa-fw fa-twitter-square" aria-hidden="true"></i> Twitter</a></li>
    
    
    
    
    
    <li><a target="_blank" rel="noopener noreferrer" href="http://localhost:4000/feed.xml"><i class="fas fa-fw fa-rss-square" aria-hidden="true"></i> Feed</a></li>
  </ul>
</div>

<div class="page__footer-copyright">
  &copy; 2023 Biogeography and Macroecology 2023. <br>

      </footer>
    </div>

    
  <script src="/assets/js/main.min.js"></script>










  </body>
</html>
